### PR TITLE
feat:[PLG-2498][PLG-2749]: Added support for ssh and winrm secrets and winrm instance name

### PR DESCRIPTION
**Pull request by devinder-harness**
Added support for ssh and winrm secrets and winrm instance name
SSH:
`harness secret --file ~/path/to/file apply --secret-type sshkey --username username --port 22`
WINRM:
`harness secret --token randompasswordtoken apply --secret-type WinRmCredentials --username username --port 5986 --domain https://abc.com`

`harness infrastructure  --file vm-aws/ssh/4-infrastructure-definition.yml apply --cloud-region us-gov-1 --instance-name ec2`

JIRA:
https://harness.atlassian.net/browse/PLG-2498
https://harness.atlassian.net/browse/PLG-2749

### CODE REVIEW

File: connector.go [hunk 1]
```diff
@@ -64,6 +68,21 @@ func applyConnector(c *cli.Context) error {
 		content = replacePlaceholderValues(content, REGION_NAME_PLACEHOLDER, awsRegion)
 		content = replacePlaceholderValues(content, DELEGATE_NAME_PLACEHOLDER, delegateName)
 	}
+	if isPdcConnectorYAML(content) {
+		hasPortNumber := strings.Contains(content, HOST_PORT_PLACEHOLDER)
+		if hostIpOrFqdn == "" || hostIpOrFqdn == HOST_IP_PLACEHOLDER {
+			githubUsername = TextInput("Enter valid host ip / fqdn:")
+		}
+		if hasPortNumber && (hostPort == "" || hostPort == HOST_PORT_PLACEHOLDER) {
+			githubUsername = TextInput("Enter valid host port:")
```
Comment (Debanitrkl) [inline connector.go:77]: here the `githubUsername` variable is being reused for different inputs (host IP/FQDN and host port), distinct variable name would be better.

Comment (devinder-harness) [reply connector.go]: it was a miss updated in recent commit

File: secret.go [hunk 2]
```diff
@@ -16,17 +17,25 @@ import (
...
 	requiresFile := isFileTypeSecret(secretType)
 	secretIdentifier := getSecretIdentifier(secretType)
 	secretName := getSecretName(secretType)
 	var secretBody HarnessSecret
-	var headers map[string]string
 	var err error
 
+	portNumber, portErr := strconv.Atoi(port)
+
+	if portErr != nil {
```
Comment (neelam-harness) [inline secret.go:36]: is this a required check or was just for troubleshooting purposes?

File: secret.go [hunk 3]
```diff
@@ -35,11 +44,9 @@ func applySecret(ctx *cli.Context) error {
 		println("Secret cannot be an empty string")
 		return nil
 	}
-	if requiresFile {
-		headers = make(map[string]string)
-		headers["Content-Type"] = "multipart/form-data"
+	if authType == "" {
+		authType = NTLM
```
Comment (neelam-harness) [inline secret.go:48]: should we add the following link in the comment for the readers to know what NTLM is ?
` // https://www.crowdstrike.com/cybersecurity-101/ntlm-windows-new-technology-lan-manager/`

File: secret.go [hunk 4]
```diff
@@ -207,3 +298,142 @@ func readSecretFromPath(filePath string, secretSpec HarnessSecret) (*bytes.Buffe
...
+
+func createSSHSecret(filepath string, secretIdentifier string, baseURL string, port int, username string, requiresFile bool) error {
+	var err error
+	var secretBody HarnessSecret
+
+	isSSHFileSecret := secretIdentifier == ""
+	if isSSHFileSecret {
+		secretIdentifier = SSH_KEY_FILE_SECRET_IDENTIFIER
+	}
+	createUrl := "v2/secrets"
```
Comment (neelam-harness) [inline secret.go:310]: Let's declare the endpoints in the constants on the similar lines to other endpoints

File: secret.go [hunk 5]
```diff
@@ -123,24 +149,89 @@ func getGitSecret(userVal string) string {
...
+		secretTypeData := WinRMSecretType{
+			Auth: WinRMSecretAuth{
+				Type: NTLM,
+				Spec: WinRMSecretSpec{
+					Username: "",
+					Password: "",
+					Domain:   "",
+				},
+			},
+			Port:       5985,
```
Comment (neelam-harness) [inline secret.go:211]: Let's not hardcode the Ports here and define them in the constants file with a descriptive name ?

**Review by neelam-harness — CHANGES_REQUESTED** [review]

File: secret.go (hunk 5 above)
Comment (devinder-harness) [reply secret.go:211]: this is just a default value we are already taking it as input

Comment (devinder-harness) [reply secret.go:211]: think of it as placeholder

File: secret.go (hunk 2 above)
Comment (devinder-harness) [reply secret.go:36]: required check as port number is a number

**Review by Ompragash — COMMENTED** [review]
- We can safely remove the below commands from the UI wizard as for PDC we don't need GitHub connectors.
```
# Create a Secret for GitHub PAT: This securely stores the token. 
harness secret  --token GITHUB-PAT apply 
# Create a GiHub Connector: Access to YAMLs. 
harness connector --file vm-pdc/winrm/github-connector.yml apply --git-user GITHUB-USERNAME 
```


- For pdc-windows cli secret creation command, it would be great if we change the flag name `--token` to `--password` as it's actually a user password
```
#  Create a Secret for WinRM: This securely stores the token. 

harness secret --token lh2fwZrHwWuT%;AFrq0J9%zRrc-ogWQ& apply --secret-type WinRmCredentials --username Administrator --domain ap-southeast-2.compute.amazonaws.com --port 5985
```
- Also, we should place Windows User Password in double quotes as it will fail as the password contains special characters in it.
`lh2fwZrHwWuT%;AFrq0J9%zRrc-ogWQ&` to `"lh2fwZrHwWuT%;AFrq0J9%zRrc-ogWQ&"`

The command fails only if we pass `--port 5985` but on running without that flag, the utility asks for user input and then Viola! pdc-connector is created.
```
harness connector --file vm-pdc/winrm/pdc-connector.yml apply --delegate-name docker-delegate --host-ip ec2-3-27-67-217.ap-southeast-2.compute.amazonaws.com --port 5985 
```

Attaching trace for reference:

```
/harness connector --file vm-pdc/winrm/pdc-connector.yml apply --delegate-name docker-delegate --host-ip ec2-3-27-67-217.ap-southeast-2.compute.amazonaws.com --port 5985
Incorrect Usage: flag provided but not defined: -port

Did you mean "--host-ip"?

NAME:
   harness connector apply - Create a new connector or Update  an existing one.

USAGE:
   harness connector apply command [command options] [arguments...]

COMMANDS:
   help, h  Shows a list of commands or help for one command

OPTIONS:
   --delegate-name value               delegate name for the k8s connector
   --git-user value                    git username for the github connector
   --aws-cross-account-role-arn value  cross account role arn for the aws connector
   --aws-access-key value              access key for the aws connector
   --aws-secret-Key value              access secret for the aws connector
   --cloud-region value                region for the cloud connector
   --host-ip value                     host ip or fqdn for the physical data center connector
   --help, -h                          show help
FATA[2023-09-07T15:53:29+05:30] flag provided but not defined: -port
```

File: secret.go (hunk 3 above)
Comment (devinder-harness) [reply secret.go:44]: @Ompragash @Debanitrkl  suggestions here ?

Comment (Debanitrkl) [reply secret.go:44]: Yeah we can add a comment

**Review by Ompragash — APPROVED** [review]
LGTM

File: secret.go (hunk 3 above)
Comment (Ompragash) [reply secret.go:44]: IMO, I don't think we need to as
1. We don't have to add an external link in the code comments (also not an official Microsoft link).
2. It'll make more sense to add somewhere in the documentation but not in the code comment section.

**Review by neelam-harness — APPROVED** [review]

**Pull request merged**